File: backend/datums.py
import requests
from pandas import DataFrame
from schema import Station


def get_stations(station_type):
    url = f'https://api.tidesandcurrents.noaa.gov/mdapi/prod/webapi/stations.json?type={station_type}&expand=tidepredoffsets&extra=details&units=english'
    response = requests.get(url)

    if not response.ok:
        return response.status_code
    
    json = response.json()
    stations = []
    for station in json['stations']:
        stations.append(Station(station['lng'], station['lat'], station['id'], resp_data=station))

    return stations

def get_datums(station_id):
    url = f'https://api.tidesandcurrents.noaa.gov/mdapi/prod/webapi/stations/{station_id}/datums.json?units=english'
    response = requests.get(url)

    if not response.ok:
        return None

    json = response.json()
    result = dict()

    for datum in json['datums']:
        result[datum['name']] = datum['value']

    datum_names = ['epoch', 'LAT', 'HAT', 'min', 'max']

    for datum_name in datum_names:
        if json[datum_name] is not None:
            result[datum_name] = json[datum_name]

    return result


def get_predictions(begin_date, end_date, datum, station, time_zone):
    url = f"https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=predictions&begin_date={begin_date}&end_date={end_date}&datum={datum}&station={station}&time_zone={time_zone}&units=english&format=json"
    response = requests.get(url)

    if not response.ok:
        return None

    return DataFrame.from_records(response.json()["predictions"])


def get_water_level(begin_date, end_date, datum, station, time_zone):
    url = f"https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=water_level&begin_date={begin_date}&end_date={end_date}&datum={datum}&station={station}&time_zone={time_zone}&units=english&format=json"
    response = requests.get(url)

    if not response.ok:
        return None

    return DataFrame.from_records(response.json()["data"])

File: backend/datums.js
const axios = require("axios");

const baseUrl = "https://api.tidesandcurrents.noaa.gov/";

async function getWaterLevel(date, time, datum, station, timeZone) {
	let url =
		baseUrl +
		`api/prod/datagetter?product=water_level&begin_date=${date}&end_date=${date}&station=${station}&datum=${datum}&time_zone=${timeZone}&units=english&format=json`;
	let resp = await axios.get(url);
	let value = resp.data.data.find((val) => val.t.split(" ")[1] == time).v;
	return {
		lat: parseFloat(resp.data.metadata.lat),
		lng: parseFloat(resp.data.metadata.lon),
		value: parseFloat(value),
	};
}

function getWaterLevels(date, time, datum, stations, timeZone) {
	let promises = [];
	for (let station of stations) {
		promises.push(getWaterLevel(date, time, datum, station, timeZone));
	}
	return Promise.all(promises);
}

async function getPrediction(date, time, datum, station, timeZone) {
	let predUrl =
		baseUrl +
		`api/prod/datagetter?product=predictions&begin_date=${date}&end_date=${date}&station=${station}&datum=${datum}&time_zone=${timeZone}&units=english&format=json`;
	let metaUrl = baseUrl + `mdapi/prod/webapi/stations/${station}.json`;
	let resp = await Promise.all([axios.get(predUrl), axios.get(metaUrl)]);
	let value = resp[0].data.predictions.find(
		(val) => val.t.split(" ")[1] == time
	).v;
	return {
		lat: resp[1].data.stations[0].lat,
		lng: resp[1].data.stations[0].lng,
		value: parseFloat(value),
	};
}

function getPredictions(date, time, datum, stations, timeZone) {
	let promises = [];
	for (let station of stations) {
		promises.push(getPrediction(date, time, datum, station, timeZone));
	}
	return Promise.all(promises);
}

module.exports = {
	getWaterLevel,
	getWaterLevels,
	getPrediction,
	getPredictions,
};
